Main.hs

```diff
@@ -337,22 +338,33 @@ processPragma qn = getUniqueCompilerPragma pragmaName qn >>= \case
     return $ ClassPragma NoCode
   Just (CompilerPragma _ s) | s == "class" ->
     return $ ClassPragma YesCode
-  _                                        -> return DefaultPragma
-
+  Just (CompilerPragma _ s) | "deriving" `isPrefixOf` s ->
+    -- parse a deriving clause for a datatype by tacking it onto a
+    -- dummy datatype and then only keeping the deriving part
+    case Hs.parseDecl ("data X = X " ++ s) of
+      Hs.ParseFailed loc msg ->
+        setCurrentRange (srcLocToRange loc) $ genericError msg
+      Hs.ParseOk m | Hs.DataDecl _ _ _ _ _ ds <- m ->
+        return $ DerivingPragma ds
+      Hs.ParseOk _ -> __IMPOSSIBLE__
+  _ -> return DefaultPragma
 compile :: Options -> ModuleEnv -> IsMain -> Definition -> TCM CompiledDef
 compile _ _ _ def = processPragma (defName def) >>= \ case
   NoPragma -> return []
   DefaultPragma -> compile' def
   ClassPragma NoCode -> return []
   ClassPragma YesCode -> return []
+  DerivingPragma ds   -> tag <$> compileData (map (() <$) ds) def
+      where tag code = [(nameBindingSite $ qnameName $ defName def, code)]
```

Comment:
I think it makes sense to merge `compile` with `compile'`, I don't see any use-sites that uses only one of them.

Reply:
I have merged `compile` and `compile'`.